### ISSUE TITLE
Fellow frustrated Python packager

### ISSUE THREAD

**Issue by Zearin** [emallson/nppm#2]
Every single time I’ve worked on a Python project, it was fun until I had to think about packaging it.  

_Every_.  _Single_.  _Time_.

I gather you feel similarly.

In searching for alternatives, I found [`pypackage`](https://github.com/ccpgames/pypackage) first, and then this project.

I prefer your project’s command line interface, and the mirroring of `npm` ’s conventions.  However, there is one thing I love about `pypackage`’s approach (excerpt taken from its README):

```
In my perfect utopian dream world, I'd see projects not have a `setup.py` 
under source control. Instead, there would only be a static metadata file. 
In the distribution version of the package, the inverse would be true.
```

In other words, developers do their stuff without worrying about a `setup.py`.  When they publish their project to PyPI, a `setup.py` is generated from the static metadata file, and that allows existing tools to do what they’ve always done.  Meanwhile, life is much easier for the developer!

**I wonder if you two could collaborate?**  (I posed a similar question in ccpgames/pypackage#5.)

I think the CLI, `package.json` conventions, and everything else in this project are the way to go.  They follow the successful (and battle-tested) conventions of NPM—and allows NPM devs to reuse existing knowledge, making it trivial for them to start up a Python project and see what’s possible.  

Dealing with `setup.py` and `setup.cfg` are a pain, though.  If NPM can get everything done with only `package.json`, `nppm` (or `ppm`) should be able to do the same.  I think ccpgames/pypackage is onto the right idea by dropping `setup.py`.

**Comment by emallson** [comment]
Hey, I think we're pretty close to the same page.

### Motivation

My motivation for this was not packaging (note the complete lack of packaging-related commands), but actually reproducible builds. A lot of code I write is sort-of one-off research code and I need to be sure that I can run it in the future. Writing and maintaining a `requirements.txt` is time-consuming and obnoxious, and I don't want to store a (technically machine-specific but practically portable) virtualenv with every project. This project basically replaces the `requirements.txt` with `package.json` to make it easy to go from source code to running the code in very little time.

I took a quick look at `pypackage` and it looks like they are much more focused on packaging and distribution. For example, `nppm` puts everything in a local `python_modules` folder while `pypackage` follows the global (or user-local) python convention. On the other hand, `pypackage` has dedicated support for testing (which I need to add at some point) and building the wheel of the package.

### setup.py

In theory, `nppm` could be modified to generate a `setup.py` directly from the `package.json`. This is one of the things I had wanted to implement. Then `setup.py` wouldn't be in VC, with the `package.json` taking its place (or, more generally, we'd have one constant `setup.py` that would read values from `package.json` so there isn't any duplication; a command like `ppm dist` would copy this file from a `site`  directory to the package directory). It is important to note that **unless the `setup.py` is in VC, the project then has an implicit dependency on nppm**. My preference would be to keep _both_ `setup.py` and `package.json` in VC, because this plays nicest with the existing python ecosystem. However, `setup.py` could easily be a static file that's the same for all `nppm`-based projects.

There are, however, a couple of things that discourage me from doing this with the project in the state its in: 
1. Complete lack of tests. I basically have a single integration test that I run locally before comitting. This ties into: 
2. Hard-dependency on pip. More than that: it uses _internal functions_ from pip that can and do change without warning (I've already dealt with that a couple of times). It also inherits the unfortunate behavior of pip when it comes to resolving packages (first version specified wins instead of error-on-conflict). This also makes it really hard to test.

Until these two things are resolved, I'm uncomfortable adding more features.

### In Summary

I would be interested in collab with @ccpgames, but we're solving fundamentally different problems. I wanted to be able to clone a repo and run it easily; they want to be able to easily build, test, and then distribute a package. I would _love_ a tool that would do both, and right now I think there is room in the python ecosystem for just that. At this point, though, these two projects have sort of orthogonal goals.

Note that I don't really have much experience with distributing python packages. The most useful thing for me would be to have someone(s) to consult with about what the _correct_ way to build+distribute packages is (because screwing that up would be very, very bad).
